File: src/mncs_validator/attestation.py
"""Offline Ed25519 DSSE-compatible attestations."""

# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

import base64
import hashlib
import os
from dataclasses import asdict, dataclass
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey

from .canonical import canonicalize, parse_json_bytes
from .errors import MncsError
from .hashing import read_regular_file

PAYLOAD_TYPE = "application/vnd.mncs.attestation-statement.v0.2+json"
ALGORITHM = "ed25519"


@dataclass(frozen=True)
class SignatureResult:
    keyid: str
    cryptographically_valid: bool
    reason: str


@dataclass(frozen=True)
class AttestationVerification:
    payload_valid: bool
    expired: bool
    signatures: list[SignatureResult]
    statement: dict[str, Any] | None

    @property
    def cryptographically_valid(self) -> bool:
        return self.payload_valid and any(item.cryptographically_valid for item in self.signatures)

    def as_dict(self) -> dict[str, Any]:
        value = asdict(self)
        value["cryptographically_valid"] = self.cryptographically_valid
        return value


def _key_id(raw_public_key: bytes) -> str:
    return f"sha256:{hashlib.sha256(raw_public_key).hexdigest()}"


def _public_record(public_key: Ed25519PublicKey) -> dict[str, Any]:
    raw = public_key.public_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PublicFormat.Raw,
    )
    return {
        "schema_version": "0.2",
        "algorithm": ALGORITHM,
        "keyid": _key_id(raw),
        "public_key": base64.b64encode(raw).decode("ascii"),
        "extensions": {},
    }


def generate_key(private_path: Path, public_path: Path | None = None) -> dict[str, Any]:
    """Create a private key only at an explicit new path and write its public record."""

    if private_path.exists():
        raise MncsError(f"refusing to overwrite private key: {private_path}")
    private_path.parent.mkdir(parents=True, exist_ok=True)
    key = Ed25519PrivateKey.generate()
    pem = key.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption(),
    )
    flags = os.O_WRONLY | os.O_CREAT | os.O_EXCL
    descriptor = os.open(private_path, flags, 0o600)
    try:
        os.write(descriptor, pem)
        os.fchmod(descriptor, 0o600)
    finally:
        os.close(descriptor)
    record = _public_record(key.public_key())
    output = public_path or private_path.with_suffix(private_path.suffix + ".pub.json")
    if output.exists():
        private_path.unlink()
        raise MncsError(f"refusing to overwrite public key record: {output}")
    output.write_bytes(canonicalize(record) + b"\n")
    return {"private_key": str(private_path), "public_key": str(output), **record}


def load_public_record(path: Path) -> dict[str, Any]:
    value = parse_json_bytes(read_regular_file(path))
    if not isinstance(value, dict):
        raise MncsError("public key record must be an object")
    if value.get("algorithm") != ALGORITHM:
        raise MncsError("unsupported key algorithm")
    raw = base64.b64decode(str(value.get("public_key", "")), validate=True)
    if len(raw) != 32 or value.get("keyid") != _key_id(raw):
        raise MncsError("public key record has an invalid or colliding key ID")
    return value


def inspect_key(path: Path) -> dict[str, Any]:
    """Inspect either a public record or an Ed25519 private key without exposing it."""

    content = read_regular_file(path)
    if content.startswith(b"-----BEGIN PRIVATE KEY-----"):
        key = serialization.load_pem_private_key(content, password=None)
        if not isinstance(key, Ed25519PrivateKey):
            raise MncsError("private key is not Ed25519")
        return {"private": True, **_public_record(key.public_key())}
    return {"private": False, **load_public_record(path)}


def _load_private(path: Path) -> Ed25519PrivateKey:
    key = serialization.load_pem_private_key(read_regular_file(path), password=None)
    if not isinstance(key, Ed25519PrivateKey):
        raise MncsError("private key is not Ed25519")
    return key


def _pae(payload_type: str, payload: bytes) -> bytes:
    type_bytes = payload_type.encode("utf-8")
    return (
        b"DSSEv1 "
        + str(len(type_bytes)).encode("ascii")
        + b" "
        + type_bytes
        + b" "
        + str(len(payload)).encode("ascii")
        + b" "
        + payload
    )


def create_statement(
    *,
    subjects: list[dict[str, Any]],
    contract_id: str,
    component: dict[str, Any],
    environment: str,
    predicate_type: str,
    predicate: dict[str, Any],
    created_at: str,
    expires_at: str | None = None,
    extensions: dict[str, Any] | None = None,
) -> dict[str, Any]:
    statement: dict[str, Any] = {
        "_type": "https://mncs.dev/attestation/v0.2/statement",
        "mncs_version": "0.2",
        "schema_version": "0.2",
        "subject": subjects,
        "contract_id": contract_id,
        "component": component,
        "environment": environment,
        "predicate_type": predicate_type,
        "predicate": predicate,
        "created_at": created_at,
        "extensions": extensions or {},
    }
    if expires_at is not None:
        statement["expires_at"] = expires_at
    return statement


def attest(
    statement: dict[str, Any], private_path: Path, envelope: dict[str, Any] | None = None
) -> dict[str, Any]:
    """Sign canonical statement bytes and append a unique signature."""

    key = _load_private(private_path)
    record = _public_record(key.public_key())
    payload = canonicalize(statement)
    encoded_payload = base64.b64encode(payload).decode("ascii")
    result = envelope or {"payloadType": PAYLOAD_TYPE, "payload": encoded_payload, "signatures": []}
    if result.get("payloadType") != PAYLOAD_TYPE or result.get("payload") != encoded_payload:
        raise MncsError("existing envelope payload does not match the statement")
    signatures = result.get("signatures")
    if not isinstance(signatures, list):
        raise MncsError("envelope signatures must be an array")
    if any(isinstance(item, dict) and item.get("keyid") == record["keyid"] for item in signatures):
        raise MncsError(f"duplicate signature from {record['keyid']}")
    signature = key.sign(_pae(PAYLOAD_TYPE, payload))
    signatures.append(
        {
            "keyid": record["keyid"],
            "algorithm": ALGORITHM,
            "sig": base64.b64encode(signature).decode("ascii"),
        }
    )
    return result


def _parse_time(value: Any) -> datetime | None:
    if not isinstance(value, str):
        return None
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
        return parsed if parsed.tzinfo is not None else None
    except ValueError:
        return None


def verify_attestation(
    envelope: dict[str, Any],
    keys: list[dict[str, Any]],
    *,
    expected_subject: str | None = None,
    expected_contract: str | None = None,
    expected_environment: str | None = None,
    now: datetime | None = None,
) -> AttestationVerification:
    """Verify DSSE signatures and bindings without applying trust policy."""

    payload_valid = envelope.get("payloadType") == PAYLOAD_TYPE
    try:
        payload = base64.b64decode(str(envelope.get("payload", "")), validate=True)
        statement_value = parse_json_bytes(payload)
        statement = statement_value if isinstance(statement_value, dict) else None
        payload_valid = (
            payload_valid and statement is not None and canonicalize(statement) == payload
        )
    except (ValueError, MncsError):
        payload = b""
        statement = None
        payload_valid = False
    expired = False
    if statement is not None:
        expires = _parse_time(statement.get("expires_at"))
        expired = expires is not None and (now or datetime.now(UTC)) >= expires
        subjects = statement.get("subject", [])
        subject_hashes = {
            digest
            for item in subjects
            if isinstance(item, dict)
            for digest in [item.get("digest", {}).get("sha256")]
            if isinstance(digest, str)
        }
        if (
            expected_subject is not None
            and expected_subject.removeprefix("sha256:") not in subject_hashes
        ):
            payload_valid = False
        if expected_contract is not None and statement.get("contract_id") != expected_contract:
            payload_valid = False
        if (
            expected_environment is not None
            and statement.get("environment") != expected_environment
        ):
            payload_valid = False
    key_map = {str(item.get("keyid")): item for item in keys}
    seen: set[str] = set()
    results: list[SignatureResult] = []
    signatures = envelope.get("signatures", [])
    if not isinstance(signatures, list):
        signatures = []
        payload_valid = False
    for item in signatures:
        if not isinstance(item, dict):
            results.append(SignatureResult("", False, "malformed signature"))
            continue
        keyid = str(item.get("keyid", ""))
        if keyid in seen:
            results.append(SignatureResult(keyid, False, "duplicate signature"))
            payload_valid = False
            continue
        seen.add(keyid)
        record = key_map.get(keyid)
        if record is None:
            results.append(SignatureResult(keyid, False, "public key unavailable"))
            continue
        try:
            if item.get("algorithm") != ALGORITHM:
                raise ValueError("algorithm confusion")
            raw = base64.b64decode(str(record["public_key"]), validate=True)
            if _key_id(raw) != keyid:
                raise ValueError("key ID mismatch")
            signature = base64.b64decode(str(item["sig"]), validate=True)
            Ed25519PublicKey.from_public_bytes(raw).verify(
                signature,
                _pae(PAYLOAD_TYPE, payload),
            )
        except (InvalidSignature, KeyError, ValueError):
            results.append(SignatureResult(keyid, False, "invalid signature"))
        else:
            results.append(SignatureResult(keyid, True, "valid signature"))
    return AttestationVerification(payload_valid, expired, results, statement)


def load_json(path: Path) -> dict[str, Any]:
    value = parse_json_bytes(read_regular_file(path))
    if not isinstance(value, dict):
        raise MncsError(f"expected JSON object: {path}")
    return value


def write_json(path: Path, value: dict[str, Any]) -> None:
    path.write_bytes(canonicalize(value) + b"\n")

File: independent/rc-consumer/src/attestation.rs
//! Offline Ed25519 verification for the MNCS DSSE-compatible envelope subset.

use std::collections::{BTreeMap, BTreeSet};

use base64::Engine;
use base64::engine::general_purpose::STANDARD;
use ed25519_dalek::{Signature, VerifyingKey};
use serde::Serialize;
use serde_json::{Map, Value};
use sha2::{Digest, Sha256};
use time::{OffsetDateTime, format_description::well_known::Rfc3339};

use crate::{canonical, json};

pub const PAYLOAD_TYPE: &str = "application/vnd.mncs.attestation-statement.v0.2+json";

#[derive(Debug, Clone, Serialize)]
pub struct SignatureResult {
    pub keyid: String,
    pub cryptographically_valid: bool,
    pub reason: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct Verification {
    pub payload_valid: bool,
    pub expired: bool,
    pub cryptographically_valid: bool,
    pub signatures: Vec<SignatureResult>,
    pub statement: Option<Value>,
    pub issues: Vec<String>,
}

fn pae(payload_type: &str, payload: &[u8]) -> Vec<u8> {
    format!(
        "DSSEv1 {} {} {} ",
        payload_type.len(),
        payload_type,
        payload.len()
    )
    .into_bytes()
    .into_iter()
    .chain(payload.iter().copied())
    .collect()
}

fn key_id(raw: &[u8]) -> String {
    format!("sha256:{}", hex::encode(Sha256::digest(raw)))
}

fn parse_time(value: Option<&Value>) -> Option<OffsetDateTime> {
    value
        .and_then(Value::as_str)
        .and_then(|text| OffsetDateTime::parse(text, &Rfc3339).ok())
}

fn public_keys(records: &[Value]) -> Result<BTreeMap<String, VerifyingKey>, String> {
    let mut keys = BTreeMap::new();
    for record in records {
        let object = record
            .as_object()
            .ok_or_else(|| "key record must be an object".to_owned())?;
        if object.get("schema_version").and_then(Value::as_str) != Some("0.2")
            || object.get("algorithm").and_then(Value::as_str) != Some("ed25519")
            || !object.get("extensions").is_some_and(Value::is_object)
        {
            return Err("unsupported or malformed public-key record".to_owned());
        }
        let keyid = object
            .get("keyid")
            .and_then(Value::as_str)
            .ok_or_else(|| "key record has no keyid".to_owned())?;
        let raw = STANDARD
            .decode(
                object
                    .get("public_key")
                    .and_then(Value::as_str)
                    .ok_or_else(|| "key record has no public_key".to_owned())?,
            )
            .map_err(|error| error.to_string())?;
        let bytes: [u8; 32] = raw
            .try_into()
            .map_err(|_| "invalid Ed25519 public-key length".to_owned())?;
        if key_id(&bytes) != keyid {
            return Err("key ID does not match public-key bytes".to_owned());
        }
        let key = VerifyingKey::from_bytes(&bytes).map_err(|error| error.to_string())?;
        if keys.insert(keyid.to_owned(), key).is_some() {
            return Err("duplicate public key ID".to_owned());
        }
    }
    Ok(keys)
}

fn subject_hashes(statement: &Map<String, Value>) -> BTreeSet<&str> {
    statement
        .get("subject")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(Value::as_object)
        .filter_map(|item| item.get("digest"))
        .filter_map(Value::as_object)
        .filter_map(|digest| digest.get("sha256"))
        .filter_map(Value::as_str)
        .collect()
}

fn valid_sha256(value: Option<&Value>) -> bool {
    value.and_then(Value::as_str).is_some_and(|text| {
        text.len() == 71
            && text.starts_with("sha256:")
            && text[7..]
                .bytes()
                .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
    })
}

fn validate_statement(statement: &Map<String, Value>, issues: &mut Vec<String>) {
    let allowed_predicates = [
        "https://mncs.dev/predicate/conformance-result/v0.2",
        "https://mncs.dev/predicate/gate-result/v0.2",
        "https://mncs.dev/predicate/evidence-index/v0.2",
        "https://mncs.dev/predicate/reproducible-package/v0.2",
        "https://mncs.dev/predicate/provider-result/v0.2",
        "https://mncs.dev/predicate/release-artifact/v0.2",
    ];
    let component = statement.get("component").and_then(Value::as_object);
    let valid = statement.get("_type").and_then(Value::as_str)
        == Some("https://mncs.dev/attestation/v0.2/statement")
        && statement.get("mncs_version").and_then(Value::as_str) == Some("0.2")
        && statement.get("schema_version").and_then(Value::as_str) == Some("0.2")
        && statement
            .get("subject")
            .and_then(Value::as_array)
            .is_some_and(|values| !values.is_empty())
        && statement
            .get("contract_id")
            .and_then(Value::as_str)
            .is_some_and(|value| !value.is_empty())
        && component.is_some_and(|value| {
            value
                .get("name")
                .and_then(Value::as_str)
                .is_some_and(|item| !item.is_empty())
                && value
                    .get("version")
                    .and_then(Value::as_str)
                    .is_some_and(|item| !item.is_empty())
                && valid_sha256(value.get("identity"))
        })
        && statement
            .get("environment")
            .and_then(Value::as_str)
            .is_some_and(|value| !value.is_empty())
        && statement
            .get("predicate_type")
            .and_then(Value::as_str)
            .is_some_and(|value| allowed_predicates.contains(&value))
        && statement.get("predicate").is_some_and(Value::is_object)
        && parse_time(statement.get("created_at")).is_some()
        && statement.get("extensions").is_some_and(Value::is_object);
    if !valid {
        issues.push("attestation statement schema violation".to_owned());
    }
    if statement.contains_key("expires_at") && parse_time(statement.get("expires_at")).is_none() {
        issues.push("invalid attestation expiration".to_owned());
    }
    for subject in statement
        .get("subject")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
    {
        let valid = subject
            .get("name")
            .and_then(Value::as_str)
            .is_some_and(|value| !value.is_empty())
            && valid_sha256(
                subject
                    .pointer("/digest/sha256")
                    .map(|value| {
                        if let Some(text) = value.as_str() {
                            Value::String(format!("sha256:{text}"))
                        } else {
                            Value::Null
                        }
                    })
                    .as_ref(),
            );
        if !valid {
            issues.push("invalid attestation subject identity".to_owned());
        }
    }
}

/// Validate an MNCS DSSE envelope and its Ed25519 signatures.
///
/// # Errors
///
/// Returns an error when an input key record cannot be decoded safely.
#[allow(clippy::too_many_lines)]
pub fn verify(
    envelope: &Value,
    key_records: &[Value],
    expected_subject: Option<&str>,
    expected_contract: Option<&str>,
    expected_environment: Option<&str>,
    now: OffsetDateTime,
) -> Result<Verification, String> {
    let object = envelope
        .as_object()
        .ok_or_else(|| "envelope must be an object".to_owned())?;
    let mut issues = Vec::new();
    if object.len() != 3
        || !object.contains_key("payloadType")
        || !object.contains_key("payload")
        || !object.contains_key("signatures")
    {
        issues.push("attestation envelope schema violation".to_owned());
    }
    let payload_type = object
        .get("payloadType")
        .and_then(Value::as_str)
        .unwrap_or_default();
    if payload_type != PAYLOAD_TYPE {
        issues.push("unsupported attestation payload type".to_owned());
    }
    let payload = object
        .get("payload")
        .and_then(Value::as_str)
        .and_then(|value| STANDARD.decode(value).ok())
        .unwrap_or_default();
    let statement = json::parse_slice(&payload).ok();
    let statement_object = statement.as_ref().and_then(Value::as_object);
    if statement_object.is_none()
        || canonical::canonicalize(statement.as_ref().unwrap_or(&Value::Null))
            .ok()
            .as_deref()
            != Some(payload.as_slice())
    {
        issues.push("attestation payload is not canonical JSON".to_owned());
    }
    if let Some(statement) = statement_object {
        validate_statement(statement, &mut issues);
        if let Some(expected) =
            expected_subject.map(|value| value.strip_prefix("sha256:").unwrap_or(value))
            && !subject_hashes(statement).contains(expected)
        {
            issues.push("attestation subject binding mismatch".to_owned());
        }
        if let Some(expected) = expected_contract
            && statement.get("contract_id").and_then(Value::as_str) != Some(expected)
        {
            issues.push("attestation contract binding mismatch".to_owned());
        }
        if let Some(expected) = expected_environment
            && statement.get("environment").and_then(Value::as_str) != Some(expected)
        {
            issues.push("attestation environment binding mismatch".to_owned());
        }
    }
    let expired = statement_object
        .and_then(|value| parse_time(value.get("expires_at")))
        .is_some_and(|expiration| now >= expiration);
    let keys = public_keys(key_records)?;
    let signatures = object
        .get("signatures")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    if signatures.is_empty() {
        issues.push("attestation signatures must be a non-empty array".to_owned());
    }
    let mut seen = BTreeSet::new();
    let mut results = Vec::new();
    for item in signatures {
        let Some(signature_object) = item.as_object() else {
            results.push(SignatureResult {
                keyid: String::new(),
                cryptographically_valid: false,
                reason: "malformed signature".to_owned(),
            });
            issues.push("malformed signature".to_owned());
            continue;
        };
        let keyid = signature_object
            .get("keyid")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_owned();
        if signature_object.len() != 3 || !seen.insert(keyid.clone()) {
            results.push(SignatureResult {
                keyid,
                cryptographically_valid: false,
                reason: "duplicate or malformed signature".to_owned(),
            });
            issues.push("duplicate or malformed signature".to_owned());
            continue;
        }
        let valid = (|| -> Result<(), String> {
            if signature_object.get("algorithm").and_then(Value::as_str) != Some("ed25519") {
                return Err("algorithm confusion".to_owned());
            }
            let key = keys
                .get(&keyid)
                .ok_or_else(|| "public key unavailable".to_owned())?;
            let raw = STANDARD
                .decode(
                    signature_object
                        .get("sig")
                        .and_then(Value::as_str)
                        .ok_or_else(|| "signature missing".to_owned())?,
                )
                .map_err(|error| error.to_string())?;
            let signature = Signature::from_slice(&raw).map_err(|error| error.to_string())?;
            key.verify_strict(&pae(payload_type, &payload), &signature)
                .map_err(|error| error.to_string())
        })()
        .is_ok();
        results.push(SignatureResult {
            keyid,
            cryptographically_valid: valid,
            reason: if valid {
                "valid signature".to_owned()
            } else {
                "invalid signature".to_owned()
            },
        });
    }
    if !results.iter().any(|item| item.cryptographically_valid) {
        issues.push("no cryptographically valid signature".to_owned());
    }
    issues.sort();
    issues.dedup();
    let payload_valid = issues
        .iter()
        .all(|issue| issue == "no cryptographically valid signature");
    let cryptographically_valid =
        payload_valid && results.iter().any(|item| item.cryptographically_valid);
    Ok(Verification {
        payload_valid,
        expired,
        cryptographically_valid,
        signatures: results,
        statement,
        issues,
    })
}
